homeassistant/components/guardian/valve.py

```diff
@@ -0,0 +1,190 @@
+
+from . import GuardianData, ValveControllerEntity, ValveControllerEntityDescription
+from .const import API_VALVE_STATUS, DOMAIN
+from .util import convert_exceptions_to_homeassistant_error
+
+ATTR_AVG_CURRENT = "average_current"
+ATTR_CONNECTED_CLIENTS = "connected_clients"
+ATTR_INST_CURRENT = "instantaneous_current"
+ATTR_INST_CURRENT_DDT = "instantaneous_current_ddt"
+ATTR_STATION_CONNECTED = "station_connected"
+ATTR_TRAVEL_COUNT = "travel_count"
+
+VALVE_KIND_VALVE = "valve"
+
+
+class GuardianValveState(StrEnum):
+    """States of a valve."""
+
+    CLOSED = "closed"
+    CLOSING = "closing"
+    FINISH_CLOSING = "finish_closing"
+    FINISH_OPENING = "finish_opening"
+    OPEN = "open"
+    OPENING = "opening"
+    START_CLOSING = "start_closing"
+    START_OPENING = "start_opening"
+
+
+@dataclass(frozen=True, kw_only=True)
+class ValveControllerValveDescription(
+    ValveEntityDescription, ValveControllerEntityDescription
+):
+    """Describe a Guardian valve controller valve."""
+
+    extra_state_attributes_fn: Callable[[dict[str, Any]], Mapping[str, Any]]
+    is_closed_fn: Callable[[dict[str, Any]], bool]
+    is_closing_fn: Callable[[dict[str, Any]], bool]
+    is_opening_fn: Callable[[dict[str, Any]], bool]
+    close_coro_fn: Callable[[Client], Coroutine[Any, Any, None]]
+    halt_coro_fn: Callable[[Client], Coroutine[Any, Any, None]]
+    open_coro_fn: Callable[[Client], Coroutine[Any, Any, None]]
+
+
+async def async_close_valve(client: Client) -> None:
+    """Close the valve."""
+    async with client:
+        await client.valve.close()
+
+
+async def async_halt_valve(client: Client) -> None:
+    """Halt the valve."""
+    async with client:
+        await client.valve.halt()
+
+
+async def async_open_valve(client: Client) -> None:
+    """Open the valve."""
+    async with client:
+        await client.valve.open()
+
+
+@callback
+def is_closing(data: dict[str, Any]) -> bool:
+    """Return if the valve is closing."""
+    return data["state"] in (
+        GuardianValveState.CLOSING,
+        GuardianValveState.FINISH_CLOSING,
+        GuardianValveState.START_CLOSING,
+    )
+
+
+@callback
+def is_opening(data: dict[str, Any]) -> bool:
+    """Return if the valve is opening."""
+    return data["state"] in (
+        GuardianValveState.OPENING,
+        GuardianValveState.FINISH_OPENING,
+        GuardianValveState.START_OPENING,
+    )
+
+
+VALVE_CONTROLLER_DESCRIPTIONS = (
+    ValveControllerValveDescription(
+        key=VALVE_KIND_VALVE,
+        translation_key="valve_controller",
+        device_class=ValveDeviceClass.WATER,
+        api_category=API_VALVE_STATUS,
+        extra_state_attributes_fn=lambda data: {
+            ATTR_AVG_CURRENT: data["average_current"],
+            ATTR_INST_CURRENT: data["instantaneous_current"],
+            ATTR_INST_CURRENT_DDT: data["instantaneous_current_ddt"],
+            ATTR_TRAVEL_COUNT: data["travel_count"],
```

Comment:
What do these measurements mean? Do you need to know if the valve is closed or opened to understand them?

Reply:
<img width="576" alt="Screenshot 2024-01-10 at 8 34 59 PM" src="https://github.com/home-assistant/core/assets/663432/e3dd82a6-1d1f-428a-9093-e61c9f135054">

Instant current only shows when the valve is in motion AFAICT

Travel count is the number of times its ever moved.

As soon I saw your comment, my first thought was I think they are probably better off as diagnostic sensors.

Reply:
Happy to make them diagnostics sensors. 👍🏻 Should I leave these attributes in place for a deprecation period, or just move them out in a breaking change right away?

Reply:
It's a new entity so it won't be breaking if we move them now before this PR is released.

Reply:
https://github.com/home-assistant/core/pull/107834